Separate CLI and NRE scopes when integrating plugins

ericnordelo opened the pull request:
Fixes #115 

- [ ] Existent plugins need to be updated (the entry points section in **pyproject.toml**)
- [ ] The [plugin template](https://github.com/franalgaba/nile-plugin-example) needs to be updated, too, to reflect this new format.

martriay commented:
@franalgaba this will break the plugin template

martriay commented on README.md:
```diff
@@ -484,7 +484,7 @@ In order for this implementation to be functional, it is needed by the plugin de
 How to decide if I want to use a plugin or not? Just install / uninstall the plugin dependency from your project :smile:
 
-Finally, after the desired plugin is installed, it will also be automatically available through the `nre`. The plugin developer should be aware of this and design the interface accordingly.
+Finally, the developer can make plugin's entry points available from nre by changing the entrypoint to `[tool.poetry.plugins.nile_plugins.nre]`. Using both cli and nre scopes in nile_plugins allows it to develop powerful plugins which are easily integrated.
```
it's confusing that in here says to change it to `nre` while the template above says "do not change this". it's also not clear how to use both. can we add an example?

martriay reviewed (commented):
could you write a small guide/doc on how to migrate a plugin?

ericnordelo commented:
@martriay I did a refactor of the README, and added an example of how to use both CLI and NRE based on your suggestion. For a guide to migrate plugins, I think it would make more sense to add it to the plugin boilerplate itself instead of our README, but anyway, I think with the example I added is enough to act as migration guide.

martriay commented on README.md:
```diff
@@ -445,46 +445,56 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
 How to decide if I want to use a plugin or not? Just install / uninstall the plugin dependency from your project :smile:
 
-Finally, after the desired plugin is installed, it will also be automatically available through the `nre`. The plugin developer should be aware of this and design the interface accordingly.
+Using both cli and nre entry points under the nile_plugins group allows it to develop powerful plugins which are easily integrated.
```
not sure what "it" is in this sentence

ericnordelo replied:
Agree and fixed!

martriay reviewed (commented):
very nice! left some final comments

andrew-fleming commented on README.md:
````diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
    ```
 
-3. Done!
+3. Optionally specify plugin entry points for `NileRuntimeEnvironment`, this doesn't require implementing a click command (remove the cli entry points if not needed):
````
```suggestion
3. Optionally specify plugin entry points for `NileRuntimeEnvironment`. This doesn't require implementing a click command (remove the cli entry points if not needed):
```

andrew-fleming commented on README.md:
````diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
+       "greet" = "nile_greet.nre.greet"
+   ```
+
+4. Done! For a better understanding of python entry points through setuptools [check this documentation](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins).
````
```suggestion
4. Done! For a better understanding of python entry points through setuptools, [check this documentation](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins).
```

andrew-fleming commented on README.md:
```diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
+In order for this implementation to be functional, it is needed by the plugin developer to follow some guidelines:
 
-1. Define a Python module that implements a click command or group:
+1. Use click if the plugin provides a CLI command:
```
What do you think about treating click as a proper noun and capitalizing it? => "Use Click if..."
Or wrapping it in backticks? => "Use `click` if..."

ericnordelo replied:
Looks better for sure.

ericnordelo replied:
Actually, I went to do the change, and I'm not so sure anymore, because we use the package name in these cases in the rest of the code (nile, nile-coverage, click, etc...). Using capitalization is a bit weird in cases like Nile-Coverage? Nile Coverage? And formatting the package name as code, doesn't look that good IMO. We should define a format for this, I think in the documentation task for the next milestone.

andrew-fleming replied:
> We should define a format for this, I think in the documentation task for the next milestone.

Yeah, good call. I think that's a better idea 👍

martriay replied:
Click [calls itself Click](https://click.palletsprojects.com/en/8.1.x/) so can do the same. I agree on a special treatment of some kind to highlight it's a proper noun.

andrew-fleming reviewed (commented):
Great work @ericnordelo! I left a few suggestions that might help with the README's readability :)

ericnordelo commented on README.md:
````diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
    ```
 
-2. Define the plugin entrypoint. In this case using Poetry features in the pyproject.toml file:
+2. Define the plugin entry points. In this case using Poetry plugins feature in the pyproject.toml file:
````
Poetry is one option, but it is not the only one. I wanted to make this implicit, but I will rewrite the sentence.

andrew-fleming replied:
Ahh I see, thank you for clarifying :)

andrew-fleming reviewed (approved):
Excellent improvements! I left one tiny suggestion. Otherwise, this looks good to me!

martriay commented on README.md:
````diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
    ```
 
-2. Define the plugin entrypoint. In this case using Poetry features in the pyproject.toml file:
+2. Define the plugin entry points (in this case by using Poetry plugins feature in the pyproject.toml file):
````
```suggestion
2. Define the plugin entry points (in this case by using Poetry plugins feature in the `pyproject.toml` file):
```

martriay commented on README.md:
````diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
    ```
 
-3. Done!
+3. Optionally specify plugin entry points for `NileRuntimeEnvironment`. This doesn't require implementing a click command (remove the cli entry points if not needed):
````
```suggestion
3. Optionally specify plugin entry points for `NileRuntimeEnvironment`. This doesn't require implementing a Click command (remove the cli entry points if not needed):
```

martriay commented on README.md:
```diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
 ### How it works
 
-This implementation takes advantage of the native extensibility features of [click](https://click.palletsprojects.com/). Using click and leveraging the Python [entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/) we have a simple manner of handling extension natively on Python environments through dependencies. The plugin implementation on Nile looks for specific Pyth ...
+This implementation takes advantage of the native extensibility features of [click](https://click.palletsprojects.com/). By using click and leveraging Python [entry points](https://packaging.python.org/en/latest/specifications/entry-points/), we have a simple manner of handling extensions natively in Python environments through dependencies. The plugin implementation on Nile looks for specific Py ...
```
```suggestion
This implementation takes advantage of the native extensibility features of [Click](https://click.palletsprojects.com/). By using Click and leveraging Python [entry points](https://packaging.python.org/en/latest/specifications/entry-points/), we have a simple manner of handling extensions natively in Python environments through dependencies. The plugin implementation on Nile looks for specific Python entry point constraints for adding commands to either the CLI or NRE.
```

martriay commented on README.md:
```diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
+In order for this implementation to be functional, the plugin developer must follow some guidelines:
 
-1. Define a Python module that implements a click command or group:
+1. Use click if the plugin provides a CLI command:
```
```suggestion
1. Use Click if the plugin provides a CLI command:
```

martriay commented on README.md:
```diff
@@ -445,46 +445,57 @@ Note that if you want to pass the token name as a hex or an int, you need to pro
-   [tool.poetry.plugins."nile_plugins"]
-   # Here you specify you command name and location <command_name> = <package_method_location>
-   "greet" = "nile_greet.main.greet"
+   # We need to specify that click commands are entry points in the group `nile_plugins`
```
```suggestion
   # We need to specify that Click commands are entry points in the group `nile_plugins`
```

andrew-fleming reviewed (approved):
Looks good to go!
